### ISSUE TITLE
[DFA-arch] Retrofit lane: DFA-adapt a FROZEN BP-pretrained gtx backbone (small-output fine-tuning; the highest-value application) — vs BP-adapt, vs frozen-adapt control

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#161]
Consuming Tao experiment: **F12** (`specs/F12-dfa-retrofit.md`, drafted). The last de-risker before the capstone; the highest-value application if positive (DFA-LoRA-class fine-tuning of models nobody can afford to full-BP).

## Why

The program's original F12 asked: can DFA retrofit an EXISTING BP-trained model — freeze the backbone, add DFA-trained capacity, continue training? **The reframe from everything since #152–#160:** a full-LM retrofit at ~50k vocab would just re-confirm the output-dim negative (F13/F16/**F22**). So F12 must live in the **small-output regime where DFA works** — and F22 (toy#160) just proved a small-output **attention** transformer IS DFA-trainable. So the sharp question:

> Take a BP-pretrained **gtx** backbone, **freeze** it, add adapter/expert capacity, continue-train on a task the frozen backbone alone **cannot** solve. Does DFA-trained added capacity (a) recover BP-trained, (b) beat a **frozen-adapter control**, and (c) do it **without a backward pass through the frozen backbone** (DFA projects the output error straight to the adapters — the cost/reach win)?

This is the DFA-LoRA / DFA-adapter thesis, now grounded in the output-dim law + F22, and it slots directly into the graph-LLM capstone (adapt a pretrained backbone with DFA).

## What to build

A **retrofit mode on the gtx lane** (reuse #160's architecture — it's the proven small-output attention testbed):

```
# phase 1 — pretrain a backbone with BP, emit a checkpoint
toy train gtx --policy chain,chain --lr 0.003 --steps N --ckpt-every K --out $BB

# phase 2 — retrofit: load, freeze backbone, add adapter capacity, continue
toy train gtx --retrofit --load-ckpt $BB/ckpt --freeze-backbone \
   --adapter-policy chain|dfa|frozen --lr <per-arm> --steps 1500 --out $DIR
```

- **Checkpoint surface for gtx:** `--ckpt-every` / `--load-ckpt` exist elsewhere (franken-moe); wire them into gtx so a BP backbone can be saved and reloaded. (You know the right shape here — a reloadable bundle the retrofit phase reads.)
- **`--freeze-backbone`:** freeze the loaded backbone weights (attention blocks + embeddings); only the ADDED adapter capacity trains. Analogous to `--freeze-experts`/`--freeze-embed`.
- **Added capacity + `--adapter-policy chain|dfa|frozen`:** the retrofit capacity — parallel adapters per block, or added blocks, your call — trained by BP (`chain`, backward through the frozen backbone), DFA (`dfa`, **direct feedback from the output error, NO backward through the backbone** — the cost win), or held frozen (`frozen`, the mandatory control). Keep DFA and BP paths CLEANLY SEPARATED (frozen backbone + separated adapters), never sum DFA into a BP-handled path — that's the F9d/#141 net-harmful shape and this design must avoid it by construction.
- **Retrofit task:** a task the frozen backbone alone **cannot** solve — a distribution shift or new relation types vs the pretrain task (so the added capacity is genuinely needed). Held fixed across arms.
- **Cost instrument:** emit the toy#159 `stream:`-style analytic figure for the **backbone-backward bytes AVOIDED** by dfa-adapt (no gradient through the frozen backbone). This is a STREAMING/structural property — analytic, stated next to any measured graph bytes, NOT a measured graph number (tao#21).

## Scope / discipline (carry the program's rules)

- **CPU-only, small output** (per the output-dim law — this is the whole point).
- **MANDATORY frozen-adapter control** (tao#19): a positive = dfa-adapt recovers bp-adapt AND clearly beats frozen-adapt.
- **control-must-be-able-to-lose (F9e/F18):** the retrofit task must be hard enough that a frozen backbone + frozen adapters clearly loses — **gate `bp-adapt >> frozen-adapt` as a precondition** before any DFA comparison means anything. This is exactly the trap #152/#154/#160 each hit; assert it.
- **PER-ARM LR (F22/#160, load-bearing):** DFA wants ~3× lower LR than BP; at BP's LR the DFA arm can read chance. The gate cell must use each arm's own best LR, or it will invert the verdict.

## Gate legs (suggested)

- **precondition:** bp-adapt val_acc − frozen-adapt val_acc > (a real margin) on the retrofit task — else the cell can't discriminate.
- `--freeze-backbone` actually freezes: backbone weights byte-identical pre/post retrofit.
- `--adapter-policy dfa` + `--dfa-b-seed` moves the adapter curve and is inert on chain (B reaches the adapters — the #157/#158 discipline).
- determinism: two identical dfa-adapt runs byte-identical.
- the analytic backbone-backward-bytes figure is emitted and is 0 for dfa-adapt vs >0 for bp-adapt.

## What this does NOT do

Not a full-LM retrofit (output-dim law says that fails — no need to re-run it). Not a claim of measured wall-clock speedup (the cost win is a streaming property, analytic per tao#21). It answers exactly one thing: **in the regime where DFA works (small output, F22), can it cheaply adapt a frozen pretrained backbone and beat the frozen control?**

## Refs
- Tao **F12** spec (drafted); **F22/toy#160** (small-output attention IS DFA-trainable — the backbone), **F10** (transfer washout — the skeptic prior that DFA can't cash in frozen pretrained features), **F9d/#141 + F18** (spine-carried artifact — why frozen-backbone + separated adapters is the right structure), **tao#19** (frozen control), **tao#21 / toy#159** (cost is analytic streaming, not measured), **toy#152** (DFA tolerates less LR). Existing surface to reuse: `--load-ckpt`, `--ckpt-every`, `--freeze-experts`, `--freeze-embed`.

**Comment by OriPekelman** [comment]
## Shipped in `31e10d1` — DFA-adapt MATCHES BP-adapt on a frozen BP-pretrained backbone

`toy train gtx --retrofit`. Phase 1 BP-pretrains the toy#160 backbone on the 16-class relation task; phase 2 **freezes and detaches** it, adds a pair-site adapter stack plus a fresh 4-class head, and trains only the added capacity under `--adapter-policy chain|dfa|frozen`.

### The result

1500 + 1500 steps, seeds 0/1/2, each arm at its own LR:

| arm | lr | s0 | s1 | s2 | mean |
|---|---|---|---|---|---|
| bp-adapt (chain) | 0.003 | .990 | .987 | .979 | .985 |
| **dfa-adapt** | 0.001 | .989 | .991 | .980 | **.987** |
| frozen-adapt | 0.003 | .313 | .314 | .295 | .307 |

**DFA-adapt recovers bp-adapt — a hair above it — and beats the frozen control by .68.** The DFA-LoRA thesis holds in the regime the output-dim law says DFA works.

A bonus finding: **DFA here is insensitive to the learning rate** — .985 / .987 / .985 across 0.0003 → 0.003, a 10× span. From scratch (toy#160) it needed ~3× less than BP, and at BP's rate it read chance. Adapting a frozen backbone through a small adapter is an *easy* regime for DFA, which is the best possible news for the application you care about.

### Two corrections to the spec, both load-bearing

**1. The retrofit task cannot be "new relation types", and this is provable.** The pretrain label `ty_a*TY + ty_b` is a BIJECTION from pairs to classes — and so is any relabeling of it, including new relation types. A retrained LINEAR head absorbs any bijection: set `u_c[a] = 1{a = a*}`, `v_c[b] = 1{b = b*}` and the correct class scores 2 while every other scores ≤ 1. So a frozen backbone plus a fresh head solves it, **the frozen-adapter control cannot lose, and the precondition leg you asked for would pass vacuously.**

The lane uses the **modular sum** `(ty_a + ty_b) mod TY` instead: many-to-one, and provably not absorbable (for TY=2 the four required inequalities sum to a contradiction; the argument runs for any TY). It needs an INTERACTION between the two endpoint representations.

That in turn **forces the adapter site**: the interaction exists only where the pair is formed, so adapters sit at the pair level, not inside the backbone stack. Node-level adapters could not express the retrofit task at all. Measured precondition: frozen-adapt .307 against chance .25, bp-adapt .985.

**2. The cost win comes from FREEZING, not from the credit rule.** With `--freeze-backbone` the backbone output is detached, so *neither* arm backpropagates through it — `bb_grad_bytes_avoided` is identical for chain and dfa (275 456 B at the lane defaults). DFA's own additional saving is the backward through the adapter stack only (32 768 B). DFA's structural cost advantage would only appear with capacity placed **below** the frozen stack's top, which this task's interaction requirement rules out.

The figure is analytic and has to be: the realized-graph counters exclude the backward extension, so a measured count cannot see an absent backward — freezing actually reads **+1 node** (the detach itself). That is the mirror of tao#21, and worth carrying into F12's framing before the cost half gets sold.

### On the checkpoint surface

Not built, deliberately. Both phases run in **one process**, which gives the property a checkpoint was meant to provide — every arm's backbone is bit-identical — *by construction* rather than by a round-trip that would itself need a gate. The runner prints `backbone: sig_pre=… sig_post=…` and the gate asserts they are identical under a freeze, so "the backbone did not move" is a measurement, not a promise that nothing was stepped. If you want the two-command workflow (pretrain once, retrofit many arms), `--ckpt-every`/`--load-ckpt` on gtx is a clean follow-on — the GGUF read/write surface is already there — but it is not needed for the result.

### Gated

`GATE PASS [gtx]` (12 legs): the retrofit precondition (`bp-adapt >> frozen-adapt`), the bit-identical freeze, the bar (dfa within .10 of bp, .40 clear of frozen), 5 retrofit fail-loud guards, and the retrofit being **inert** when unasked — the plain toy#160 curve is byte-identical.

That last one earned its keep: toy#160's byte fixture caught two silent bugs during this build — an adapter zero-init that matched `.up` and was zeroing every block's FFN up-projection (`b<i>.up`), and a phase flag that fed mod-sum labels to the 16-class head in *non*-retrofit runs. Neither was visible in any retrofit number.

**Comment by OriPekelman** [comment]
Closing: shipped in `31e10d1` — feat(dfa-arch): toy#161/F12 DFA retrofit — DFA-adapt MATCHES BP-adapt on a frozen BP-pretrained backbone

Verified by `git log`; the implementing commit names this issue in its subject. Part of the 2026-08-14 issue-state sweep (48 closed, all against a commit in `main`). Reopen if the scope here outran what landed.
